### PR TITLE
Text splitter

### PR DESCRIPTION
**Pull request by NirantK**

### CODE REVIEW

File: fastembed/embedding.py [hunk 1]
```diff
@@ -647,6 +683,19 @@ def embed(
...
+
+        Args:
+            text (str): The text to split.
+            chunk_size (Optional[int], optional): Maximum size of chunks based on the tokenizer encoding.
+            chunk_overlap (Optional[int], optional): Allowed overlap in characters between chunks.
+
+        Returns:
+            List[str]: The list of strings.
+        """
+        return self.model.split_text(text, chunk_size=chunk_size, chunk_overlap=chunk_overlap)
```
Comment (NirantK) [inline fastembed/embedding.py:697]: Looks like we're repeating this wrapper function across classes. Would you want to add this to the abstract class upstream from which other classes can inherit and get access to this?

File: fastembed/splitter.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        strip_whitespace: bool = True,
+    ) -> None:
+        """Create a new TextSplitter.
+
+        Args:
+            chunk_size: Maximum size of chunks to return
+            chunk_overlap: Overlap in characters between chunks
+            length_function: Function that measures the length of given chunks
+            keep_separator: Whether to keep the separator in the chunks
+            strip_whitespace: If `True`, strips whitespace from the start and end of
```
Comment (NirantK) [inline fastembed/splitter.py:54]: We mention and measure size and overlap in character or tokens? We should maintain consistency across the codebase please?

File: tests/test_splitter.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+            continue
+
+        p = Path(__file__).with_name("state_of_the_union.txt")
+        with open(p) as f:
+            text = f.read()
+            embedding = DefaultEmbedding(
+                model_name=model_desc["model"],
+                chunk_size=chunk_size,
+                chunk_overlap=chunk_overlap,
+            )
```
Comment (NirantK) [inline tests/test_splitter.py:22]: The `chunk_size` should be in tokens and not characters. And if the user passes a value larger than the Embedding model token sequence length, we should raise a ValueError.

File: fastembed/splitter.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    Recursively tries to split by different characters with
+    the tokenizer encoding to restrict length.
+    """
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        chunk_size: int = 4000,
+        chunk_overlap: int = 200,
```
Comment (NirantK) [inline fastembed/splitter.py:135]: Same challenge, we should pick tokens and these look like char values?

File: tests/test_splitter.py (hunk 3 above)
Comment (Anush008) [reply tests/test_splitter.py:22]: The chunk size is calculated based on the tokenizers encoding.

https://github.com/qdrant/fastembed/blob/text-splitter/fastembed%2Fsplitter.py#L98

File: fastembed/embedding.py (hunk 1 above)
Comment (Anush008) [reply fastembed/embedding.py]: https://github.com/qdrant/fastembed/pull/80/commits/0641f6fee0348f2e1dfb6b2d108e74795a582c5d

**Comment by NirantK** [comment]
Fix the Windows and MacOS tests please? Feel free to use the Github Actions by commit-pushing frequently if you don't have access to all 3 OS!

File: fastembed/splitter.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                    final_chunks.append(s)
+                else:
+                    other_info = self._split_text(s, new_separators)
+                    final_chunks.extend(other_info)
+        if _good_splits:
+            merged_text = self._merge_splits(_good_splits, _separator)
+            final_chunks.extend(merged_text)
+        return final_chunks
+
+    def split_text(self, text: str, chunk_size: Optional[int] = None, chunk_overlap: Optional[int] = None) -> List[str]:
```
Comment (NirantK) [inline fastembed/splitter.py:205]: I'm not quite sure why we need this wrapper function? Let's just rename the _split_text to split_text?

File: fastembed/splitter.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        chunk_size: int = 4000,
+        chunk_overlap: int = 200,
+        strip_whitespace: bool = True,
+        separators: Optional[List[str]] = None,
+        keep_separator: bool = True,
+        is_separator_regex: bool = False,
+    ) -> None:
+        """Create a new TextSplitter."""
+
+        if not isinstance(tokenizer, Tokenizer):
```
Comment (NirantK) [inline fastembed/splitter.py:143]: We should check that the maximum chunk_size is never more than the max length allowed by the tokenizer. E.g. 512 tokens for BAAI, which can be 2K for some other Tokenizer. Please add that to the docstring and ValueError on that as well.

File: tests/test_splitter.py (hunk 3 above)
Comment (NirantK) [reply tests/test_splitter.py:23]: Yes. I've added a comment as well. Please read the max tokenzier lenght from the tokenizer object. Not the config file.

Comment (NirantK) [reply tests/test_splitter.py]: Never mind, I've added this to the validator. Please add a test for this i.e. that the error is indeed raised — and then mark this conversation as resolved.

File: fastembed/splitter.py (hunk 6 above)
Comment (NirantK) [reply fastembed/splitter.py:131]: Never mind, I've added this to the validator. Please add a test for this i.e. that the error is indeed raised — and then mark this conversation as resolved — this is the same as previous comment. Can resolve both together.

File: fastembed/embedding.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 from fastembed.parallel_processor import ParallelWorkerPool, Worker
 from fastembed.splitter import FastEmbedRecursiveSplitter
 
-from .types import TextSplitterConfig
+from .models import TextSplitterConfig
```
Comment (NirantK) [inline fastembed/embedding.py:21]: Much better name! thanks!

**Comment by Anush008** [comment]
@NirantK, Got this mypy warning.
```console
"fastembed/types.py" shadows library module "types".
```
Renamed it to "models".

File: fastembed/models.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        keep_separator (bool): Whether to keep the separator or not. Default is False.
+        strip_whitespace (bool): Whether to strip whitespace or not. Default is True.
+        tokenizer (Optional[Tokenizer]): Tokenizer used in FastEmbedRecursiveSplitter. Default is None.
+        is_separator_regex (bool): Whether the separator is a regular expression. Default is False.
+        separators (Optional[List[str]]): List of separators. Default is None.
+    """
+    model_config: ConfigDict = ConfigDict(arbitrary_types_allowed=True)
+
+    chunk_size: Optional[int] = None
+    chunk_overlap: Optional[int] = None
```
Comment (Anush008) [inline fastembed/models.py:28]: @NirantK, arithmetic comparison between possible `None` values at L46, L48.

**Comment by NirantK** [comment]
Design Requests for the Text Splitter @Anush008 
---  

1. How can we tokenize once and use the same for embedding? E.g. from the recursive tokenizer, we get the tokenizer output which then gets embedded directly. This is important from a compute lens — we don't want to tokenize the same text twice — once for chunking and again for embedding. 

2. Retain document to chunk mapping. If each Point is assigned an id, use that to create a new payload called "document_id" against each chunk which comes from that document. Assign uuid4 ids to documents, if not already passed by the user. 

Please resolve the merge conflicts when you touch this and start a new PR as well. I'm closing this PR, since I raised this for convenience initially.

**Comment by Anush008** [comment]
This expansion in scope will intertwine the splitter with the rest of the FastEmbed. Since the original proposition for the splitter was to be a separate, optional, QoL improvement API.
Not sure if this will be beneficial too.

**Comment by Anush008** [comment]
> 1. How can we tokenize once and use the same for embedding? E.g. from the recursive tokenizer, we get the tokenizer output which then gets embedded directly. This is important from a compute lens — we don't want to tokenize the same text twice — once for chunking and again for embedding.

This expansion in scope will intertwine the splitter with the rest of the FastEmbed. Since the original proposition for the splitter was to be a separate, optional, QoL improvement API.
Not sure if this will be beneficial too.

> 2. Retain document to chunk mapping. If each Point is assigned an id, use that to create a new payload called "document_id" against each chunk which comes from that document. Assign uuid4 ids to documents, if not already passed by the user.

Pertains to and to be addressed in `qdrant_client` I assume.

**Comment by NirantK** [comment]
Re #1: Why would a single use tokenizer force intertwine with the rest of the FastEmbed? We anyway create tokens before we embed. This changes the forking point between this and the current implementation — because we move around tokens instead of strings now. 

Re: #2 We'll have to add metadata to the load/response from FastEmbed to Qdrant Client — that is the change I'm requesting here. We'll have to make corresponding changes in the Qdrant Client as well, yes.

**Comment by Anush008** [comment]
The text splitter method is expected to return list of split texts correct? Which users can then chain to the embedding methods.

**Pull request closed without merging**